lib/hanami/application.rb

```diff
@@ -1,64 +1,282 @@
 # frozen_string_literal: true
 
+require "dry/monitor"
+require "dry/monitor/rack/middleware"
+require "dry/system/container"
+require "dry/system/components"
 require "hanami/configuration"
-require "hanami/routes"
-require "hanami/router"
+require "pathname"
+require "rack"
+require_relative "slice"
+require_relative "application/router"
+require_relative "web/rack_logger"
 
 module Hanami
-  # Hanami application
-  #
-  # @since 2.0.0
   class Application
     @_mutex = Mutex.new
 
-    def self.inherited(app_class)
-      @_mutex.synchronize do
-        app_class.class_eval do
-          @_mutex         = Mutex.new
-          @_configuration = Hanami::Configuration.new(env: Hanami.env)
+    class << self
+      def inherited(klass)
+        @_mutex.synchronize do
+          klass.class_eval do
+            @_mutex         = Mutex.new
+            @_configuration = Hanami::Configuration.new(env: Hanami.env)
 
-          extend ClassMethods
-          include InstanceMethods
-        end
+            extend ClassMethods
+            include InstanceMethods
+          end
 
-        Hanami.application_class = app_class
+          Hanami.application = klass
+        end
       end
     end
 
-    # Class method interface
-    #
-    # @since 2.0.0
     module ClassMethods
       def configuration
         @_configuration
       end
 
-      alias config configuration
+      alias_method :config, :configuration
+
+      def init
+        return self if inited?
+
+        configuration.finalize
+
+        @container = prepare_container
+        @deps_module = prepare_deps_module
+        @slices = load_slices
+        @inited = true
+        self
+      end
+
+      def container
+        raise "Application not init'ed" unless defined?(@container)
+        @container
+      end
+
+      def deps
+        raise "Application not init'ed" unless defined?(@deps_module)
+        @deps_module
+      end
+
+      def slices
+        raise "Application not init'ed" unless defined?(@slices)
+        @slices
+      end
+
+      def register_bootable(*args, &block)
+        container.boot(*args, &block)
+      end
+
+      def init_bootable(*args)
+        container.init(*args)
+      end
+
+      def start_bootable(*args)
+        container.start(*args)
+      end
+
+      def keys
+        container.keys
+      end
+
+      def [](*args)
+        container.[](*args)
+      end
+
+      def resolve(*args)
+        container.resolve(*args)
+      end
+
+      def boot(&block)
+        return self if booted?
+
+        init
+
+        container.configure do; end # force after configure hook
```

Comment:
I think we should introduce proper setup callbacks (like in rom) instead of relying on `configure`. I wanted to do this anyway so I wonder what your thoughts are.

Reply:
Agreed, having to do this clearly indicates that some real lifecycle hooks are missing.

Reply:
Noted here: https://github.com/dry-rb/dry-system/issues/129